### PR TITLE
Fix BasicPersistor handling of heavy fields

### PR DESCRIPTION
**Pull request by valentin-woven**
Context: BasicPersistor does not handle heavy pointers correctly. For these fields, the binary data is saved in the parquet file (as well as the columns files) rather than the location information.

What changed:
1. Added a local storage abstraction, to replace FakeStorage in situation when we want to load an S3Datataset from local storage. Since the dataset loads the parquet file directly, the storage abstraction is bypassed and a LocalFileSystem instance needs to be used instead.
2. Changed the test_persistence.py file to include loading the data via an S3Dataset object in order to highlight the problem.
3. Fixed the logic in BasicPersistor. Also removed the partitioning / chunking logic, as it was not consistent with what the Spark persistor is doing.

### CODE REVIEW

**Comment by isaak-willett** [comment]
@valentin-woven did you find that whole heavy pointers are being saved in the pyarrow table? Can you point me to a dataset?

**Comment by isaak-willett** [comment]
Testing plan:
1. Create dataset with current Wicker version on Main
2. Release this as an rc version
3. Create dataset with rc version
4. Compare the pyarrow tables
5. Compare the rows of the datasets for equivalency

**Comment by isaak-willett** [comment]
@valentin-woven I'm planning on fully reviewing this tomorrow morning before 8am PST, does that work for you? I've done most of the work of getting the datasets I just had something come up.

**Comment by valentin-woven** [comment]
> @valentin-woven I'm planning on fully reviewing this tomorrow morning before 8am PST, does that work for you? I've done most of the work of getting the datasets I just had something come up.

Sure, that works, thank you very much @isaak-willett !

**Review by isaak-willett — APPROVED** [review]
I'm going to put the testing plan after this approval but the fix does as it denotes!
v0.0.4 = 162650012 bytes
v0.0.5a = 4100012 bytest

**Review by isaak-willett — APPROVED** [review]
I'm going to put the testing plan after this approval but the fix does as it denotes!
v0.0.4 = 162650012 bytes
v0.0.5a = 4100012 bytest

**Comment by isaak-willett** [comment]
I ran the following scripts to verify, these make the following given changing the version of wicker
1. A cifar10 dataset with wicker 0.0.4
2. A cifar10 dataset with wicker 0.0.5a

```python
from wicker import schema

DATASET_SCHEMA = schema.DatasetSchema(
    fields=[
        schema.StringField("label", description="ground truth label of our image"),
        schema.StringField("id", description="uniqe id of data sample"),
        schema.NumpyField("image", shape=(-1, -1, 3), dtype="uint8", description="image as a numpy array"),
    ],
    primary_keys=["label", "id"], # tuple must be unique over the dataset
)

DATASET_NAME = "cifar10_wicker_v5a"
DATASET_VERSION = "0.0.1"

import os
import numpy as np
import uuid

from PIL import Image

# replace with default path to cifar10
DEFAULT_DATA_PATH = '/home/ubuntu/code/wicker/cifar10' #os.getcwd()+'/cifar10'

if not os.path.isdir(DEFAULT_DATA_PATH):
    raise OSError(f"cifar10 directory does not exist at {DEFAULT_DATA_PATH}")

path = os.walk(DEFAULT_DATA_PATH)
dataset = []
for root, directories, files in path:
    if files:
        root_split = root.split('/')
        partition = root_split[-2] # train, test folders
        label = root_split[-1] # images for each label type are in a folder with that label name
        print(f'{partition}, {label}')
    for file in files:
        im = np.asarray(Image.open(root+'/'+file))
        # Tuple(partition, sample data dictionary)
        dataset.append((partition, 
                        {'label': label, 'id': str(uuid.uuid4()), 'image': im}))
        
from wicker.core.persistance import BasicPersistor
import copy

persistor = BasicPersistor()
persistor.persist_wicker_dataset(
    DATASET_NAME,
    DATASET_VERSION,
    DATASET_SCHEMA,
    dataset,
)
```

```python
from wicker.core.datasets import S3Dataset

DATASET_NAME_4 = "cifar10_wicker_v4"
DATASET_NAME_5 = "cifar10_wicker_v5a"
DATASET_VERSION = "0.0.1"

ds_4 = S3Dataset(DATASET_NAME_4, DATASET_VERSION, "train", columns_to_load=['label', 'image'])
ds_5 = S3Dataset(DATASET_NAME_5, DATASET_VERSION, "train", columns_to_load=['label', 'image'])
ds_4_pyarrow_table = ds_4.arrow_table()
ds_5_pyarrow_table = ds_5.arrow_table()
print(ds_4_pyarrow_table.nbytes)
print(ds_5_pyarrow_table.nbytes)

row_4_0 = ds_4[1]
row_5_0 = ds_5[1]
print(row_4_0)
print(row_5_0)

```

**Comment by isaak-willett** [comment]
@valentin-woven I bumped the version and did a release to v0.0.5, I'll let you merge this into main though. @varshaparthay Im done with the review on this

File: wicker/core/persistance.py
Comment (valentin-woven) [reply wicker/core/persistance.py:236]: The partitions that the spark persistor is making are driven by the data (which comes in tuples [partition, row_data]) rather than the persistor chunking it.
In practice, our workflows will actually have data for one partition only ("train" or "eval") so there is a missed opportunity there. The train/eval should be separate datasets and the persistors should have flexibility to partition the dataset but that's for another day.

**Comment by valentin-woven** [comment]
Many thanks for your help @isaak-willett !

**Pull request merged**
